include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+            stream_buf->overflow(); // ensures that put area has space available
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator++()
+    {
+        return *this;
+    }
+    //!\overload
+    fast_ostreambuf_iterator & operator++(int)
+    {
+        return *this;
+    }
+    //!\}
+
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator*()
+    {
+        return *this;
+    }
+
+    //!\brief Writes a character to the associated output stream.
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            stream_buf->sputc(c); // overflow() [virtual], then write character
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes exactly `n` characters `[it, it + n)` to the associated output.
+     * \tparam iterator_type The iterator type; Must model std::ranges::input_iterator.
+     * \param[in, out] it The iterator to the range to write.
+     */
+    template <std::ranges::input_iterator iterator_type>
+    void write_n(iterator_type & it, size_t n)
+    {
+        if (size_t buffer_space = stream_buf->epptr() - stream_buf->pptr(); n <= buffer_space)
+        {
+            it = std::ranges::copy_n(it, n, stream_buf->pptr()).in; // copy_n returns incremented iterators {in, out}
+            stream_buf->pbump(n);
+        }
+        else
+        {
+            it = std::ranges::copy_n(it, buffer_space, stream_buf->pptr()).in;
+            stream_buf->pbump(buffer_space);
+
+            if (stream_buf->overflow(*it) == traits_t::eof()) // update put area by trying to write the next char
+                throw std::ios_base::failure{"Could not write to output stream."};
+
+            ++it;
+            ++buffer_space;
+
+            write_n(it, n - buffer_space);
+        }
+    }
+
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     *
+     * \note This function calls std::ranges::distance(rng) and might be inefficient for ranges that aren't sized.
+     */
+    template <std::ranges::forward_range range_type>
+    void write_range(range_type && rng)
+    {
+        auto it = rng.begin();
+        write_n(it, std::ranges::distance(rng));
+    }
```

Comment:
Why do we need two functions (write_n and write_range)?

Reply:
Convenience?
(`write_n` is more generic and good for chunked output like fasta-line-wrapping without using `view::take` all the time. But write_range is more convinient in general)

Reply:
I think, I would have designed it as a single function since view_take is free for random access ranges and even single iterators can be turned into such a range via `views::counted`, but that's up to you!

Reply:
Well but as you said earlier (that distance() might be expensive for non sized views) we do not always have random access ranges, so I thought this is more efficient ?

Reply:
It's hard to say without benchmarking to be honest.

Reply:
So I did a benchmark and the time was the same so I removed write_n now

Reply:
:+1: